ostd/src/sync/rcu/non_null.rs

```diff
@@ -0,0 +1,242 @@
+    type Target = Box<T>;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: A `Box<T>` is guaranteed to be represented by a single pointer [1] and a shared
+        // reference to the `Box<T>` during the lifetime `'a` can be created according to the
+        // safety requirements of `NonNullPtr::raw_as_ref`.
+        //
+        // [1]: https://doc.rust-lang.org/std/boxed/#memory-layout
+        unsafe { core::mem::transmute(&self.inner) }
+    }
+}
+
+impl<'a, T> BoxRef<'a, T> {
+    /// Dereferences `self` to get a reference to `T` with the lifetime `'a`.
+    pub fn deref_target(&self) -> &'a T {
+        // SAFETY: The reference is created through `NonNullPtr::raw_as_ref`, hence
+        // the original owned pointer and target must outlive the lifetime parameter `'a`,
+        // and during `'a` no mutable references to the pointer will exist.
+        unsafe { &*(self.inner) }
+    }
+}
+
+unsafe impl<T: Send + 'static> NonNullPtr for Box<T> {
+    type Target = T;
+
+    type Ref<'a>
+        = BoxRef<'a, T>
+    where
+        Self: 'a;
+
+    fn into_raw(self) -> NonNull<Self::Target> {
+        let ptr = Box::into_raw(self).cast();
+
+        // SAFETY: The pointer representing a `Box` can never be NULL.
+        unsafe { NonNull::new_unchecked(ptr) }
+    }
+
+    unsafe fn from_raw(ptr: NonNull<Self::Target>) -> Self {
+        let ptr = ptr.as_ptr().cast();
+
+        // SAFETY: The safety is upheld by the caller.
+        unsafe { Box::from_raw(ptr) }
+    }
+
+    unsafe fn raw_as_ref<'a>(raw: NonNull<Self::Target>) -> Self::Ref<'a> {
+        BoxRef {
+            inner: raw.as_ptr().cast(),
+            _marker: PhantomData,
+        }
+    }
+
+    fn ref_as_raw(ptr_ref: Self::Ref<'_>) -> NonNull<Self::Target> {
+        // SAFETY: The pointer representing a `Box` can never be NULL.
+        unsafe { NonNull::new_unchecked(ptr_ref.inner.cast()) }
+    }
+}
+
+/// A type that represents `&'a Arc<T>`.
+#[derive(Debug)]
+pub struct ArcRef<'a, T> {
+    inner: ManuallyDrop<Arc<T>>,
+    _marker: PhantomData<&'a Arc<T>>,
+}
+
+impl<T> Deref for ArcRef<'_, T> {
+    type Target = Arc<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<'a, T> ArcRef<'a, T> {
+    /// Dereferences `self` to get a reference to `T` with the lifetime `'a`.
+    pub fn deref_target(&self) -> &'a T {
+        // SAFETY: The reference is created through `NonNullPtr::raw_as_ref`, hence
+        // the original owned pointer and target must outlive the lifetime parameter `'a`,
+        // and during `'a` no mutable references to the pointer will exist.
+        unsafe { &*(self.deref().deref() as *const T) }
+    }
+}
+
+unsafe impl<T: Send + Sync + 'static> NonNullPtr for Arc<T> {
+    type Target = T;
+
+    type Ref<'a>
+        = ArcRef<'a, T>
+    where
+        Self: 'a;
+
+    fn into_raw(self) -> NonNull<Self::Target> {
+        let ptr = Arc::into_raw(self).cast_mut().cast();
```

Comment:
Oh, given that we have a reasonable `Target`. You should also drop all `cast()` now. There are many unnecessary `cast()`s in this file.